### PR TITLE
Support role assignment across subscriptions

### LINKED ISSUE #1343: Backup role assignments fail when applied across subscriptions
## :white_check_mark: Checklist

- [x] I have searched open and closed issues for duplicates.
- [x] This is a problem observed when **deploying** a Data Safe Haven.
- [x] I can reproduce this with the [latest version](https://github.com/alan-turing-institute/data-safe-haven/releases).
- [x] I have read through the [documentation](https://alan-turing-institute.github.io/data-safe-haven/).
- [x] This isn't an open-ended question (open a [discussion](https://github.com/alan-turing-institute/data-safe-haven/discussions) if it is).

## :computer: System information

- **Operating System:** macOS
- **Data Safe Haven version:** 4.0.1

## :no_entry_sign: Describe the problem

When running `Setup_SRE_Backup.ps1` the script fails while attempting to add backup roles to various SRE resources.

The root issue here is that the backup script isn't working in the case where the backup role is in one subscription and the target is it applied to is in another one. The error occurs when running Setup_SRE_Backup.ps1.

The relevant function is New-AzRoleAssignment. Note that we may also need to switch between the SHM and SRE subscriptions before/after running Deploy-RoleAssignment (which is the function in AzureResources.psm1 that applies roles to resources).

## :deciduous_tree: Log messages

<details>
<summary>Relevant log messages</summary>

```none
PS /Users/echalstrey/projects/data-safe-haven/deployment/secure_research_environment/setup> ./Setup_SRE_Backup.ps1 -shmId prod4 -sreId cefas
2022-11-25 16:52:35 [   INFO]: Ensuring that resource group 'RG_SHM_PROD4_SRE_CEFAS_BACKUP' exists...
2022-11-25 16:52:35 [   INFO]: [ ] Creating resource group 'RG_SHM_PROD4_SRE_CEFAS_BACKUP'
2022-11-25 16:52:36 [SUCCESS]: [✔] Created resource group 'RG_SHM_PROD4_SRE_CEFAS_BACKUP'
2022-11-25 16:52:36 [   INFO]: Ensuring that backup vault 'bv-prod4-sre-cefas' exists...
2022-11-25 16:52:36 [   INFO]: [ ] Creating backup vault 'bv-prod4-sre-cefas'
2022-11-25 16:52:37 [SUCCESS]: [✔] Successfully deployed backup vault bv-prod4-sre-cefas
2022-11-25 16:52:37 [   INFO]: Ensuring backup policy 'blobbackuppolicy' exists
2022-11-25 16:52:37 [   INFO]: [ ] Creating backup policy 'blobbackuppolicy'
2022-11-25 16:52:37 [SUCCESS]: [✔] Successfully deployed backup policy blobbackuppolicy

TerminatingError(Deploy-RoleAssignment): "Cannot bind argument to parameter 'ObjectId' because it is an empty string."

Deploy-RoleAssignment: /Users/echalstrey/projects/data-safe-haven/deployment/secure_research_environment/setup/Setup_SRE_Backup.ps1:50
Line |
  50 |  $null = Deploy-RoleAssignment -ObjectId $Vault.IdentityPrincipalId `
     |                                          ~~~~~~~~~~~~~~~~~~~~~~~~~~
     | Cannot bind argument to parameter 'ObjectId' because it is an empty string.

Deploy-RoleAssignment: /Users/echalstrey/projects/data-safe-haven/deployment/secure_research_environment/setup/Setup_SRE_Backup.ps1:50
Line |
  50 |  $null = Deploy-RoleAssignment -ObjectId $Vault.IdentityPrincipalId `
     |                                          ~~~~~~~~~~~~~~~~~~~~~~~~~~
     | Cannot bind argument to parameter 'ObjectId' because it is an empty string.

>> TerminatingError(Deploy-DataProtectionBackupInstance): "Cannot bind argument to parameter 'BackupPolicyId' because it is an empty string."

Deploy-DataProtectionBackupInstance: /Users/echalstrey/projects/data-safe-haven/deployment/secure_research_environment/setup/Setup_SRE_Backup.ps1:59
Line |
  59 |  … ll = Deploy-DataProtectionBackupInstance -BackupPolicyId $Policy.Id `
     |                                                             ~~~~~~~~~~
     | Cannot bind argument to parameter 'BackupPolicyId' because it is an empty string.

Deploy-DataProtectionBackupInstance: /Users/echalstrey/projects/data-safe-haven/deployment/secure_research_environment/setup/Setup_SRE_Backup.ps1:59
Line |
  59 |  … ll = Deploy-DataProtectionBackupInstance -BackupPolicyId $Policy.Id `
     |                                                             ~~~~~~~~~~
     | Cannot bind argument to parameter 'BackupPolicyId' because it is an empty string.

2022-11-25 16:52:39 [   INFO]: Ensuring backup policy 'diskbackuppolicy' exists
2022-11-25 16:52:39 [   INFO]: [ ] Creating backup policy 'diskbackuppolicy'

New-AzDataProtectionBackupPolicy_Create: /Users/echalstrey/.local/share/powershell/Modules/Az.DataProtection/0.4.0/custom/Cmdlets/Policy/New-AzDataProtectionBackupPolicy.ps1:123
Line |
 123 |          Az.DataProtection.Internal\New-AzDataProtectionBackupPolicy @ …
     |          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | The Resource 'Microsoft.DataProtection/BackupVaults/bv-prod4-sre-cefas' under resource group 'RG_SHM_PROD4_SRE_CEFAS_BACKUP' was not found. For more details please go to https://aka.ms/ARMResourceNotFoundFix

New-AzDataProtectionBackupPolicy_Create: /Users/echalstrey/.local/share/powershell/Modules/Az.DataProtection/0.4.0/custom/Cmdlets/Policy/New-AzDataProtectionBackupPolicy.ps1:123
Line |
 123 |          Az.DataProtection.Internal\New-AzDataProtectionBackupPolicy @ …
     |          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | The Resource 'Microsoft.DataProtection/BackupVaults/bv-prod4-sre-cefas' under resource group 'RG_SHM_PROD4_SRE_CEFAS_BACKUP' was not found. For more
     | details please go to https://aka.ms/ARMResourceNotFoundFix

2022-11-25 16:52:40 [SUCCESS]: [✔] Successfully deployed backup policy diskbackuppolicy
PS /Users/echalstrey/projects/data-safe-haven/deployment/secure_research_environment/setup> TerminatingError(Deploy-RoleAssignment): "Cannot bind argument to parameter 'ObjectId' because it is an empty string."

Deploy-RoleAssignment: /Users/echalstrey/projects/data-safe-haven/deployment/secure_research_environment/setup/Setup_SRE_Backup.ps1:80
Line |
  80 |  $null = Deploy-RoleAssignment -ObjectId $Vault.IdentityPrincipalId `
     |                                          ~~~~~~~~~~~~~~~~~~~~~~~~~~
     | Cannot bind argument to parameter 'ObjectId' because it is an empty string.

Deploy-RoleAssignment: /Users/echalstrey/projects/data-safe-haven/deployment/secure_research_environment/setup/Setup_SRE_Backup.ps1:80
Line |
  80 |  $null = Deploy-RoleAssignment -ObjectId $Vault.IdentityPrincipalId `
     |                                          ~~~~~~~~~~~~~~~~~~~~~~~~~~
     | Cannot bind argument to parameter 'ObjectId' because it is an empty string.

>> TerminatingError(Deploy-RoleAssignment): "Cannot bind argument to parameter 'ObjectId' because it is an empty string."

Deploy-RoleAssignment: /Users/echalstrey/projects/data-safe-haven/deployment/secure_research_environment/setup/Setup_SRE_Backup.ps1:89
Line |
  89 |  …  $null = Deploy-RoleAssignment -ObjectId $Vault.IdentityPrincipalId `
     |                                             ~~~~~~~~~~~~~~~~~~~~~~~~~~
     | Cannot bind argument to parameter 'ObjectId' because it is an empty string.

Deploy-RoleAssignment: /Users/echalstrey/projects/data-safe-haven/deployment/secure_research_environment/setup/Setup_SRE_Backup.ps1:89
Line |
  89 |  …  $null = Deploy-RoleAssignment -ObjectId $Vault.IdentityPrincipalId `
     |                                             ~~~~~~~~~~~~~~~~~~~~~~~~~~
     | Cannot bind argument to parameter 'ObjectId' because it is an empty string.

>> TerminatingError(Deploy-RoleAssignment): "Cannot bind argument to parameter 'ObjectId' because it is an empty string."

Deploy-RoleAssignment: /Users/echalstrey/projects/data-safe-haven/deployment/secure_research_environment/setup/Setup_SRE_Backup.ps1:94
Line |
  94 |  …  $null = Deploy-RoleAssignment -ObjectId $Vault.IdentityPrincipalId `
     |                                             ~~~~~~~~~~~~~~~~~~~~~~~~~~
     | Cannot bind argument to parameter 'ObjectId' because it is an empty string.

Deploy-RoleAssignment: /Users/echalstrey/projects/data-safe-haven/deployment/secure_research_environment/setup/Setup_SRE_Backup.ps1:94
Line |
  94 |  …  $null = Deploy-RoleAssignment -ObjectId $Vault.IdentityPrincipalId `
     |                                             ~~~~~~~~~~~~~~~~~~~~~~~~~~
     | Cannot bind argument to parameter 'ObjectId' because it is an empty string.

>> TerminatingError(Deploy-RoleAssignment): "Cannot bind argument to parameter 'ObjectId' because it is an empty string."

Deploy-RoleAssignment: /Users/echalstrey/projects/data-safe-haven/deployment/secure_research_environment/setup/Setup_SRE_Backup.ps1:89
Line |
  89 |  …  $null = Deploy-RoleAssignment -ObjectId $Vault.IdentityPrincipalId `
     |                                             ~~~~~~~~~~~~~~~~~~~~~~~~~~
     | Cannot bind argument to parameter 'ObjectId' because it is an empty string.

Deploy-RoleAssignment: /Users/echalstrey/projects/data-safe-haven/deployment/secure_research_environment/setup/Setup_SRE_Backup.ps1:89
Line |
  89 |  …  $null = Deploy-RoleAssignment -ObjectId $Vault.IdentityPrincipalId `
     |                                             ~~~~~~~~~~~~~~~~~~~~~~~~~~
     | Cannot bind argument to parameter 'ObjectId' because it is an empty string.

>> TerminatingError(Deploy-RoleAssignment): "Cannot bind argument to parameter 'ObjectId' because it is an empty string."

Deploy-RoleAssignment: /Users/echalstrey/projects/data-safe-haven/deployment/secure_research_environment/setup/Setup_SRE_Backup.ps1:94
Line |
  94 |  …  $null = Deploy-RoleAssignment -ObjectId $Vault.IdentityPrincipalId `
     |                                             ~~~~~~~~~~~~~~~~~~~~~~~~~~
     | Cannot bind argument to parameter 'ObjectId' because it is an empty string.

Deploy-RoleAssignment: /Users/echalstrey/projects/data-safe-haven/deployment/secure_research_environment/setup/Setup_SRE_Backup.ps1:94
Line |
  94 |  …  $null = Deploy-RoleAssignment -ObjectId $Vault.IdentityPrincipalId `
     |                                             ~~~~~~~~~~~~~~~~~~~~~~~~~~
     | Cannot bind argument to parameter 'ObjectId' because it is an empty string.

Deploy-DataProtectionBackupInstance: /Users/echalstrey/projects/data-safe-haven/deployment/secure_research_environment/setup/Setup_SRE_Backup.ps1:105
Line |
 105 |  … ll = Deploy-DataProtectionBackupInstance -BackupPolicyId $Policy.Id `
     |                                                             ~~~~~~~~~~
     | Cannot bind argument to parameter 'BackupPolicyId' because it is an empty string.

Deploy-DataProtectionBackupInstance: /Users/echalstrey/projects/data-safe-haven/deployment/secure_research_environment/setup/Setup_SRE_Backup.ps1:105
Line |
 105 |  … ll = Deploy-DataProtectionBackupInstance -BackupPolicyId $Policy.Id `
     |                                                             ~~~~~~~~~~
     | Cannot bind argument to parameter 'BackupPolicyId' because it is an empty string.

>> TerminatingError(Deploy-DataProtectionBackupInstance): "Cannot bind argument to parameter 'BackupPolicyId' because it is an empty string."

Deploy-DataProtectionBackupInstance: /Users/echalstrey/projects/data-safe-haven/deployment/secure_research_environment/setup/Setup_SRE_Backup.ps1:105
Line |
 105 |  … ll = Deploy-DataProtectionBackupInstance -BackupPolicyId $Policy.Id `
     |                                                             ~~~~~~~~~~
     | Cannot bind argument to parameter 'BackupPolicyId' because it is an empty string.

Deploy-DataProtectionBackupInstance: /Users/echalstrey/projects/data-safe-haven/deployment/secure_research_environment/setup/Setup_SRE_Backup.ps1:105
Line |
 105 |  … ll = Deploy-DataProtectionBackupInstance -BackupPolicyId $Policy.Id `
     |                                                             ~~~~~~~~~~
     | Cannot bind argument to parameter 'BackupPolicyId' because it is an empty string.

>> TerminatingError(Deploy-DataProtectionBackupInstance): "Cannot bind argument to parameter 'BackupPolicyId' because it is an empty string."

Deploy-DataProtectionBackupInstance: /Users/echalstrey/projects/data-safe-haven/deployment/secure_research_environment/setup/Setup_SRE_Backup.ps1:105
Line |
 105 |  … ll = Deploy-DataProtectionBackupInstance -BackupPolicyId $Policy.Id `
     |                                                             ~~~~~~~~~~
     | Cannot bind argument to parameter 'BackupPolicyId' because it is an empty string.

Deploy-DataProtectionBackupInstance: /Users/echalstrey/projects/data-safe-haven/deployment/secure_research_environment/setup/Setup_SRE_Backup.ps1:105
Line |
 105 |  … ll = Deploy-DataProtectionBackupInstance -BackupPolicyId $Policy.Id `
     |                                                             ~~~~~~~~~~
     | Cannot bind argument to parameter 'BackupPolicyId' because it is an empty string.
```

</details>

## :recycle: To reproduce

Deploy an SRE in a different subscription to the SHM.

## Workaround

Deploy SHM and SRE into the same Azure subscription.

### PR DESCRIPTION
**Pull request by JimMadge**
## :white_check_mark: Checklist

- [x] You have given your pull request a meaningful title (_e.g._ `Enable foobar integration` rather than `515 foobar`).
- [x] You are targeting the **develop branch**.
- [x] Your branch is up-to-date with the **develop branch** (you probably started your branch from `develop` but it may have changed since then).
- [x] If-and-only-if your changes are not yet ready to merge, you have marked this pull request as a **draft** pull request and added '[WIP]' to the title.
- [ ] If-and-only-if you have changed any Powershell code, you have run the code formatter. You can do this with `./tests/AutoFormat_Powershell.ps1 -TargetPath <path to file or directory>`.

### :arrow_heading_up: Summary

- Add option to create role assignment by scope
- Use scope for role assignment across SHM-SRE boundary

### :closed_umbrella: Related issues

Closes #1343

### CODE REVIEW

File: deployment/secure_research_environment/setup/Setup_SRE_Backup.ps1 [hunk 1]
```diff
@@ -60,9 +60,7 @@ $null = Set-AzContext -SubscriptionId $config.sre.subscriptionName -ErrorAction
 # Assign permissions required for blob backup to the Vault's managed identity
 # ---------------------------------------------------------------------------
 $null = Deploy-RoleAssignment -ObjectId $Vault.IdentityPrincipalId `
-                              -ResourceGroupName $PersistentStorageAccount.ResourceGroupName `
-                              -ResourceType "Microsoft.Storage/storageAccounts" `
-                              -ResourceName $PersistentStorageAccount.StorageAccountName `
+                              -Scope $PersistentStorageAccount.id
```
Comment (jemrobinson) [inline deployment/secure_research_environment/setup/Setup_SRE_Backup.ps1]:
```suggestion
                              -Scope $PersistentStorageAccount.id `
```

**Comment by jemrobinson** [comment]
@edwardchalstrey1 @JimMadge has this been tested on an SHM/SRE in different subscriptions?

**Comment by edwardchalstrey1** [comment]
> @edwardchalstrey1 @JimMadge has this been tested on an SHM/SRE in different subscriptions?

for some reason the backtick fix wasn't working for me:

```
PS /Users/echalstrey/projects/data-safe-haven/deployment/secure_research_environment/setup> git log --name-status HEAD^..HEAD
commit e6fb846eb8bdf7df330f0904cc9fbbbf05c09aa5
Author: James Robinson <james.em.robinson@gmail.com>
Date:   Tue Nov 29 12:05:10 2022 +0000

    :bug: Add missing backtick

M	deployment/secure_research_environment/setup/Setup_SRE_Backup.ps1
PS /Users/echalstrey/projects/data-safe-haven/deployment/secure_research_environment/setup> ./Setup_SRE_Backup.ps1 -shmId prod4 -sreId cefas
2022-11-30 13:10:59 [   INFO]: Ensuring that resource group 'RG_SHM_PROD4_SRE_CEFAS_BACKUP' exists...
2022-11-30 13:10:59 [SUCCESS]: [✔] Resource group 'RG_SHM_PROD4_SRE_CEFAS_BACKUP' already exists
2022-11-30 13:11:07 [   INFO]: Ensuring that backup vault 'bv-prod4-sre-cefas' exists...
2022-11-30 13:11:08 [SUCCESS]: [✔] Backup vault 'bv-prod4-sre-cefas' already exists
2022-11-30 13:11:08 [   INFO]: Ensuring backup policy 'blobbackuppolicy' exists
2022-11-30 13:11:08 [SUCCESS]: [✔] Backup policy 'blobbackuppolicy' already exists
cmdlet Deploy-RoleAssignment at command pipeline position 1
Supply values for the following parameters:
(Type !? for Help.)
ResourceGroupName:
```

**Comment by edwardchalstrey1** [comment]
This shows that one of the role assignments uses the SHM subscription, but another fails because of permissions lacking for the SRE subscription:

```
022-11-30 14:06:07 [SUCCESS]: [✔] Role assignment 'Storage Account Backup Contributor' over '/subscriptions/4aea9c2f-9b6c-42e8-8b09-3594994fe238/resourceGroups/RG_SHM_PROD4_PERSISTENT_DATA/providers/Microsoft.Storage/storageAccounts/prod4cefasdatakveorxfnql' already exists
2022-11-30 14:06:07 [   INFO]: Ensuring backup instance for 'prod4cefasdatakveorxfnql' exists
2022-11-30 14:06:08 [SUCCESS]: [✔] Backup instance for 'prod4cefasdatakveorxfnql' already exists
2022-11-30 14:06:08 [   INFO]: Ensuring backup policy 'diskbackuppolicy' exists
2022-11-30 14:06:08 [SUCCESS]: [✔] Backup policy 'diskbackuppolicy' already exists
2022-11-30 14:06:08 [   INFO]: Ensuring that role assignment for 3283b51b-d4fa-4c5b-bc35-8f05c107cbad as Disk Snapshot Contributor over RG_SHM_PROD4_SRE_CEFAS_BACKUP exists...
2022-11-30 14:06:09 [FAILURE]: [x] Failed to create role assignment 'Disk Snapshot Contributor' over 'RG_SHM_PROD4_SRE_CEFAS_BACKUP'
OperationStopped: Operation returned an invalid status code 'Forbidden'
```

The role I have is "contributor" for the SRE subscription - **so SRE deployers may need to be "owners" of SRE subscriptions**

Screenshot 2022-11-30 at 14 26 07

**Comment by JimMadge** [comment]
@jemrobinson looks like the issue is the account deploying here has the "Contributor" role. In our development testing we use "Owner" role. Does the documentation say anything about what Azure role you need?

**Comment by jemrobinson** [comment]
SRE deployers need to be in a group that is an "Owner" of the subscription being deployed into. Is this not the case here?

**Comment by edwardchalstrey1** [comment]
I will test this when I have the Owner access

**Comment by edwardchalstrey1** [comment]
@JimMadge tested and works

**Review by jemrobinson — APPROVED** [review]
LGTM

**Pull request merged**